Feat: implement spanner CDC for Bento

anicoll opened the pull request:
~TODO: add integration tests.~ Added

anicoll commented on go.mod:
```diff
@@ -398,7 +405,9 @@ require (
 )
 
-go 1.23.0
+go 1.24
+
+toolchain go1.24.0
```
If this is a problem, happy to undo and push a change to keep 1.23 if required

anicoll commented on Makefile:
```diff
@@ -12,7 +12,7 @@ PATHINSTSERVERLESS = $(DEST_DIR)/serverless
 DOCKER_IMAGE       ?= ghcr.io/warpstreamlabs/bento
 
-VERSION   := $(shell git describe --tags || echo "v0.0.0")
+VERSION   := $(shell git describe --tags 2>/dev/null || echo "v0.0.0")
```
handles failures gracefully (some reason happening locally) 🤷

anicoll replied:
fair enough. I will revert this

jem-davies commented:
Hi @anicoll - thanks for the contribution! 🚀  - I should (or @gregfurman) get around to taking a look in a few days. Looks like there is just one lint check failing atm - is this something you could take a look at 🙏

anicoll commented:
> Hi @anicoll - thanks for the contribution! 🚀 - I should (or @gregfurman) get around to taking a look in a few days. Looks like there is just one lint check failing atm - is this something you could take a look at 🙏

I have fixed the lint issue and added some extra metadata tests to the READ func. Also made those variables.

jem-davies reviewed (changes requested):
Pretty much just doc changes / nits on this review, I have taken a look at the implementation and mostly looks good but want to check again regarding the use of contexts - how does this handle shutdown gracefully etc.  

Thanks

anicoll commented on internal/impl/gcp/input_spanner_cdc.go:
```diff
@@ -0,0 +1,309 @@
+	cdcFieldEndTime              = "end_time"
+	cdcFieldHeartbeatInterval    = "heartbeat_interval"
+
+	metadataTimestamp   = "gcp_spanner_commit_timestamp"
+	metadataModType     = "gcp_spanner_cdc_mod_type"
+	metadataTableName   = "gcp_spanner_table_name"
```
Reason I never did _cdc on some of the fields was because I didnt feel they were cdc specific.
Such as a table name doesnt pertain to the CDC aprt. nor does the commit timestamp which is just a database thing rather than a CDC thing.

Will await for your preference before making any changes

jem-davies replied:
keep it as it is i think yeah

jem-davies reviewed (changes requested):
I think that there is a problem with gracefully shutting down this component? 

I set up a `gcr.io/cloud-spanner-emulator/emulator` similar to the integration test then manually sent SIGTERM to the bento process and got the logs in the log.txt file I have attached. 

[logs.txt](https://github.com/user-attachments/files/20618342/logs.txt)

Also I think that the connect logic doesn't retry because I did get this once as well: 
```
go run main.go -c ./tmp/0605_gcp_spanner_cdc_test/config.yaml
INFO Running main config from specified file       @service=bento bento_version="" path=./tmp/0605_gcp_spanner_cdc_test/config.yaml
INFO Listening for HTTP requests at: http://0.0.0.0:4195  @service=bento
ERRO Service closing due to: failed to init input <no label> path root.input: rpc error: code = FailedPrecondition desc = Schema change operation rejected because a concurrent schema change operation or read-write transaction is already in progress.  @service=bento
exit status 1
```

I think normally we put most of the connection logic in the Connect function of the component then the bento engine will retry to connect indefinitely rather than terminate? 

There is also a small fix needed for the tests at the moment.

anicoll commented:
@jem-davies 
I have fixed said tests (apologies) and please see attached screenshot of the handling of a sigterm now.
<img width="1173" alt="image" src="https://github.com/user-attachments/assets/9c274705-f72f-41cb-98b9-69f978d91660" />

I tested it like so
`
func main() {
	ctx := context.Background()

	go func() {
		time.Sleep(15 * time.Second)
		p, err := os.FindProcess(os.Getpid())
		if err != nil {
			panic(err)
		}
		if err := p.Signal(syscall.SIGTERM); err != nil {
			panic(err)
		}
	}()

	service.RunCLI(ctx)
}
`

Any more feedback is welcome..

anicoll commented:
This PR resolves #340

anicoll commented:
@jem-davies ping

jem-davies commented:
> @jem-davies ping

👋  - haven't forgotten about - just working through a few other things on the todo list then will come back to this 

If it's a blocker for you - I would suggest looking at implementing as a plugin.

jem-davies commented on internal/impl/gcp/input_spanner_cdc.go:
```diff
@@ -0,0 +1,329 @@
+	subCtx, cancel := context.WithCancel(context.Background())
+	c.closeFunc = cancel
+
+	go func() {
+		rerr := c.subscriber.Subscribe(subCtx, c.consumer)
+
+		var apiErr *apierror.APIError
+		if errors.As(rerr, &apiErr) {
+			if apiErr.GRPCStatus().Code() == codes.Canceled {
+				c.log.Infof("Subscription cancelled: %v\n", apiErr)
+			} else {
+				c.log.Errorf("API error during subscription: %v\n", apiErr)
+			}
+		} else {
+			c.log.Errorf("Subscription error: %v\n", rerr)
+		}
+
+		c.cdcMut.Lock()
+		close(c.consumer.msgQueue)
+		c.closeFunc = nil
+		c.cdcMut.Unlock()
+	}()
```
We need to return the errors from the connect function - so if the .Subscribe fails we should be returning that error. 
The issue I am finding is that the .Subscribe func doesn't seem to fail even if there is no chance it will connect i.e. 

Incorrectly setting the env var: `SPANNER_EMULATOR_HOST=localhost:9015` it will just hang: 

```
INFO Running main config from specified file       @service=bento bento_version="" path=./tmp/0807_gcp_spanner_cdc/config.yaml
INFO Listening for HTTP requests at: http://0.0.0.0:4195  @service=bento
``` 

or if I cause a connection error by setting the wrong dsn - I am getting panics: 

```
ERRO API error during subscription: rpc error: code = NotFound desc = Database not found: projects/project-id/instances/instance-id/databases/stream-databases
error details: name = ResourceInfo type = type.googleapis.com/google.spanner.admin.database.v1.Database resourcename = projects/project-id/instances/instance-id/databases/stream-databases owner =  desc = Database does not exist.  @service=bento label="" path=root.input
panic: Subscriber has already started subscribe.
```

What would be handy is if we had a way of testing the connection prior to calling subscribe - like the `cypher` input: 

```go

	if err := driver.VerifyConnectivity(ctx); err != nil {
		return err
	}
	if err := driver.VerifyAuthentication(ctx, nil); err != nil {
		return err
	}
```

we then return an error from Connect then Bento will go into a retry loop trying to reconnect.

jem-davies commented on internal/impl/gcp/integration_spanner_cdc_test.go:
```diff
@@ -0,0 +1,343 @@
+		_, _, err = input.Read(readCtx)
+
+		// Verify we got a context cancellation error
+		s.ErrorIs(err, service.ErrNotConnected, "Expected context.Canceled error, got: %v", err)
```
I think that this should be a context.Cancelled error not a ErrNotConnected

jem-davies reviewed (changes requested):
As Greg pointed out re. the shutdown - I think that there is changes to be made regarding initial connection as it doesn't currently follow bento semantics as is

gregfurman commented on internal/impl/gcp/input_spanner_cdc.go:
```diff
@@ -0,0 +1,419 @@
+func spannerCdcSpec() *service.ConfigSpec {
+	return service.NewConfigSpec().
+		Beta().
+		Version("1.8.0").
```
```suggestion
		Version("1.11.0").
```

jem-davies commented on internal/impl/gcp/input_spanner_cdc.go:
```diff
@@ -0,0 +1,419 @@
+	}
+
+	var err error
+	c.streamClient, err = spanner.NewClient(ctx, c.conf.SpannerDSN)
```
We will need to test this client actually connects and if not, return a service.ErrNotConnected otherwise Bento will just hang after reporting:

```
INFO Input type gcp_spanner_cdc is now active      @service=bento label="" path=root.input
```

gregfurman replied:
The `NewClient` should be creating a grpc connection to the DSN. If it can't connect, because it's unauthenticated or whatever, then this `NewClient` will fail. 

Are you suggesting we do a `healthcheck` or equivalent after connecting to gurantee that the client is _actually_ able to connect?

jem-davies commented on internal/impl/gcp/input_spanner_cdc.go:
```diff
@@ -0,0 +1,419 @@
+		conf:            conf,
+		log:             res.Logger(),
+		shutdownSig:     shutdown.NewSignaller(),
+		recordsCh:       make(chan changeRecord),
```
hmm thinking that a buffered channel would be better?

gregfurman replied:
OK sure. How large would we want it? If the channel is full, I assume we then just skip new reads?

jem-davies replied:
hmm not sure 100-1000 - we could expose via the config?

gregfurman replied:
Changed to buffered channel 😄 Thanks!

jem-davies reviewed (dismissed):
Think that the shutdown of the stream isn't handled properly, getting the following: 

ERRO Error in readPartition: spanner: code = "Canceled", desc = "context canceled", requestID = "1.bce9f9cad806f2d4.1.4.7.1"  @service=bento label="" path=root.input stream=config

and: 

```
2025/08/25 16:33:26 Number of sessions in use is negative, resetting it to currSessionsCheckedOutLocked. Stack trace: goroutine 123 [running]:
```

Edit - I did implement shutdown on: https://github.com/jem-davies/bento/tree/feat/spanner_cdc_jem (but that was for the old client)
